src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala

```diff
@@ -0,0 +1,303 @@
+    ("nHomRef", TInt),
+    ("nHet", TInt),
+    ("nHomVar", TInt),
+    ("nMissing", TInt))
+
+  def apply(vds: VariantDataset,
+    kinshipFiltExprVA: String,
+    assocFiltExprVA: String,
+    pathVA: List[String],
+    rootGA: String,
+    completeSamples: IndexedSeq[String],
+    covNames: List[String],
+    C: DenseMatrix[Double],
+    y: DenseVector[Double],
+    optDelta: Option[Double] = None,
+    noAssoc: Boolean = false,
+    useML: Boolean = false,
+    sparsityThreshold: Double = 1d,
+    useBlock: Boolean = true): VariantDataset = {
+
+    val n = y.length
+
+    val completeSamplesSet = completeSamples.toSet
+    val sampleMask = vds.sampleIds.map(completeSamplesSet).toArray
+
+    info(s"lmmreg: Computing RRM for $n samples...")
+
+    // FIXME: persist here?
+
+    val vdsKinship = vds.filterVariantsExpr(kinshipFiltExprVA, keep = true).filterSamples((s, sa) => completeSamplesSet(s))
+
+    val (rrm, m) =
+      if (useBlock)
+        ComputeRRM.withBlock(vdsKinship)
+      else
+        ComputeRRM.withoutBlock(vdsKinship)
+
+    assert(rrm.rows == n && rrm.cols == n)
+
+    info(s"lmmreg: RRM computed using $m variants")
+    info(s"lmmreg: Computing eigenvectors of RRM...")
+
+    val eigK = eigSymD(rrm)
+    val Ut = eigK.eigenvectors.t
+    val S = eigK.eigenvalues // increasing order
+    assert(S.length == n)
+
+    info("lmmreg: 20 largest evals: " + ((n - 1) to math.max(0, n - 20) by -1).map(S(_).formatted("%.5f")).mkString(", "))
+    info("lmmreg: 20 smallest evals: " + (0 until math.min(n, 20)).map(S(_).formatted("%.5f")).mkString(", "))
+
+    optDelta match {
+      case Some(_) => info(s"lmmreg: Delta specified by user")
+      case None => info(s"lmmreg: Estimating delta using ${ if (useML) "ML" else "REML" }... ")
+    }
+
+    val UtC = Ut * C
+    val Uty = Ut * y
+
+    val diagLMM = DiagLMM(UtC, Uty, S, optDelta, useML)
+
+    val delta = diagLMM.delta
+    val globalBetaMap = covNames.zip(diagLMM.globalB.toArray).toMap
+    val globalSg2 = diagLMM.globalS2
+    val globalSe2 = delta * globalSg2
+    val h2 = globalSg2 / (globalSg2 + globalSe2)
+
+    val header = "rank\teval"
+    val evalString = (0 until n).map(i => s"$i\t${ S(n - i - 1) }").mkString("\n")
+    log.info(s"\nlmmreg: table of eigenvalues\n$header\n$evalString\n")
+
+    info(s"lmmreg: global model fit: beta = $globalBetaMap")
+    info(s"lmmreg: global model fit: sigmaG2 = $globalSg2")
+    info(s"lmmreg: global model fit: sigmaE2 = $globalSe2")
+    info(s"lmmreg: global model fit: delta = $delta")
+    info(s"lmmreg: global model fit: h2 = $h2")
+    info(s"lmmreg: Computing LMM statistics for each variant...")
+
+    val T = Ut(::, *) :* diagLMM.sqrtInvD
+    val Qt = qr.reduced.justQ(diagLMM.TC).t
+    val QtTy = Qt * diagLMM.Ty
+    val TyQtTy = (diagLMM.Ty dot diagLMM.Ty) - (QtTy dot QtTy)
+
+    val assocVariantFilterPred = vds.filterVariantsExprPred(assocFiltExprVA, keep = true)
+    val keepAll = assocFiltExprVA == "true"
```

Comment:
Because in practice many users don't care about LOCO, so they do filter the vds first and don't need any addition assocFilter.  In that case, I was trying to make things a bit more efficient by not bothering to evaluate the condition each time (probably over-optimizing).  This will go away once vds is pre-filtered.